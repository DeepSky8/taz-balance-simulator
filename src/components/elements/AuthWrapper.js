import React from "react";

export const AuthWrapper = ({children}) => {

    return (
        <div>
            <div>Login</div>
            {children}
        </div>
    )
}

export { AuthWrapper as default }