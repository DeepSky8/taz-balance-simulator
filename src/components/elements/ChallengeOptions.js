import React, { useContext, useEffect, useState } from "react";
import { useReducer } from "react";
// import PickChallengeObject from "./Challenges/PickChallengeObject";
import { villainFlavorArray, villainObjectsArray, villainTitleArray } from "../mission-elements/m-villain";
import { setVillainObject } from "../../actions/challengeSelectActions";
import challengeTransformer from "../../reducers/challengeTransformer";
import PickChallengeObject from './Challenges/PickChallengeObject';

const ChallengeOptions = (
    {
        challengeObjectArray,
        challengeDispatch
    }
) => {

    const onClick = (challengeCode) => {
        // When the button for one of the 
        const challengeObject =
            challengeTransformer(challengeObjectArray, challengeCode)

        challengeDispatch(challengeObject)
    }

    return (
        <div>
            {challengeObjectArray.map((challenge) => {
                return <PickChallengeObject
                    key={challenge.challengeCode}
                    selectChallenge={() => { onClick(challenge.challengeCode) }}
                    {...challenge}
                />
            })}

        </div>
    )

}

export { ChallengeOptions as default }

// <button onClick={selectVillain}>Set villain</button>