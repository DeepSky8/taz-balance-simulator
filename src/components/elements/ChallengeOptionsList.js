import React from "react";
import challengeTransformer from "../../reducers/challengeTransformer";
import PickChallengeObject from './Challenges/PickChallengeObject';

const ChallengeOptionsList = (
    {
        challengeObjectArray,
        challengeDispatch,
        joining
    }
) => (
    <div>
        {challengeObjectArray.map((challenge) => {
            return <PickChallengeObject
                key={challenge.challengeCode}
                selectChallenge={() => { challengeDispatch(challenge) }}
                joining={joining}
                {...challenge}
            />
        })}
    </div>
)

export { ChallengeOptionsList as default }