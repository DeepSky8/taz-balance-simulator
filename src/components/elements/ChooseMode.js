import React from "react";
import { Link } from "react-router-dom";

const ChooseMode = () => {

    return (
        <div>
            <div><Link to='/gameSetup'>Join a game without an account</Link></div>
            <div><Link to='/signIn'>Sign in to host or join a game</Link></div>

        </div>
    )
}

export { ChooseMode as default }