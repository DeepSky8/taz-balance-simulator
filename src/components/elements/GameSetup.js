import React from "react";

export const GameSetup = () => {
    return (
        <p>Game Setup page</p>
    )

}

export { GameSetup as default }