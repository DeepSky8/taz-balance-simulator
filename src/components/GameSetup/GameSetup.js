import React, { useEffect, useState } from "react";
import { auth, db } from "../../firebase/firebase";
import { off, onValue, ref, } from "firebase/database";
import { clearPlayerList, updateGameState, updatePlayerList, } from "../../actions/gameActions";
import { startRemoveGameID } from "../../actions/userActions";
import AuthWrapper from "../Authentication/AuthWrapper";
import JoiningHosting from "./JoiningHosting";
import ChallengeDisplay from "../elements/Challenges/ChallengeDisplay";
import { Outlet } from "react-router-dom";
import CharacterChallengeNavBar from "./CharacterChallengeNavBar";
import PlayingAs from "../elements/Party/partyMembers/PlayingAs";

export const GameSetup = ({ dispatchGameState, userState, gameState, charState, setSavedGameArray }) => {
    const [gameArray, setGameArray] = useState([])


    // Get list of active games
    useEffect(() => {
        // Listen to list of activeGame objects in Firebase
        onValue(ref(db, 'activeGames'), (snapshot) => {

            const updatedArray = [];
            if (snapshot.exists()) {

                snapshot.forEach((childSnapShot) => {
                    updatedArray.push(childSnapShot.val().gameID)
                })
            }
            // Set a new list of current game codes on GameSetup state
            // when the listener perceives a change
            setGameArray(updatedArray)

        })

        return () => {
            // Remove the listener on Active Games in the cloud
            off(ref(db, 'activeGames'))
        }
    }, [])

    // Join Active Game
    useEffect(() => {
        if (userState.gameID) {
            // When the local gameID changes, if the local gameID exists
            // start a listener to get the game information
            onValue(ref(db, 'activeGames/' + userState.gameID), (snapshot) => {
                if (snapshot.exists()) {
                    dispatchGameState(updateGameState(snapshot.val()))
                    // dispatchGameState(updateChallengesObject(snapshot.val()))
                }
            })
            // and then start a listener to get the participating players.
            onValue(ref(db, 'activeGames/' + userState.gameID + '/playerList'), snapshot => {

                if (snapshot.exists()) {
                    const playerList = [];
                    snapshot.forEach((player) => { playerList.push(player.val()) })
                    const otherPlayers = playerList.filter(player => player.uid !== auth.currentUser.uid)
                    dispatchGameState(updatePlayerList(otherPlayers))
                }

            })
        }

        // If the Active Game record no longer exists, 
        // get rid of the stored gameID and clear the local player list

        // else {
        //     console.log('would have started StartRemoveGameID with auth: ', auth.currentUser.uid)
        //     dispatchGameState(clearPlayerList())
        //     off(ref(db, 'activeGames/' + userState.gameID))
        //     off(ref(db, 'activeGames/' + userState.gameID + '/playerList'))
        // }

        return () => {
            off(ref(db, 'activeGames/' + userState.gameID))
            off(ref(db, 'activeGames/' + userState.gameID + '/playerList'))
        }
    }, [userState.gameID])

    // Get list of saved games for this user
    useEffect(() => {
        onValue(ref(db, 'savedGames/' + auth.currentUser.uid), (snapshot) => {
            const userSavedGames = [];
            if (snapshot.exists()) {
                snapshot.forEach((savedGame) => { userSavedGames.push(savedGame.val()) })
            }
            setSavedGameArray(userSavedGames)

        })

        return () => {
            off(ref(db, 'savedGames/' + auth.currentUser.uid))
        }
    }, [])


    return (
        <div>
            <AuthWrapper />
            <JoiningHosting
                userState={userState}
                dispatchGameState={dispatchGameState}
                gameArray={gameArray}
            />
            <ChallengeDisplay
                gameState={gameState}
            />
            <PlayingAs
                userState={userState}
                charState={charState}
            />
            <CharacterChallengeNavBar

            />
            <Outlet />
            Start Game button goes here
        </div>
    )

}

export { GameSetup as default }



