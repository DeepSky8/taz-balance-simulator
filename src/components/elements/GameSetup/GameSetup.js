import React, { useEffect, useReducer, useState } from "react";
import { auth, db } from "../../../firebase/firebase";
import { off, onValue, ref } from "firebase/database";
import { Route, Routes, useNavigate } from "react-router-dom";
import {
    clearGameState,
    clearClassList,
    clearPlayerList,
    startJoinActiveGame,
    updateClassList,
    updatePlayerList,
    updateReadyList,
    updateReadyStatus,
    clearReadyList,
    startRemoveGameCode,
    updateGameStatic,
    updateGameActive,
} from "../../../actions/gameActions";
import CharacterChallengeNavBar from "./CharacterChallengeNavBar";
import PlayingAs from "./PlayingAs";
import {
    clearUserState,
    startRecordCurrentGame,
    updateUserState
} from "../../../actions/userActions";
import CharacterSelect from "./CharacterSelect/CharacterSelect";
import { setCharState, setNoCurrentChar } from "../../../actions/charActions";
import { charReducer, defaultCharState } from "../../../reducers/charReducer";
import { defaultGameState, gameReducer } from "../../../reducers/gameReducer";
import { defaultUserProfile, userReducer } from "../../../reducers/userReducer";
import NewLoadWrapper from "../Challenges/NewLoadWrapper";
import RestOfParty from "./Party/RestOfParty";
import JoiningHosting from "./JoiningHosting";
import StartGame from "./StartGame";
import GameInstructions from "./GameInstructions";
import AuthWrapper from "../../Authentication/AuthWrapper";
import ChallengeDisplay from "./ChallengeDisplay";




export const GameSetup = ({ }) => {
    let navigate = useNavigate()
    const [userState, dispatchUserState] = useReducer(userReducer, defaultUserProfile)
    const [gameArray, setGameArray] = useState([])
    const [charArray, setCharArray] = useState([])
    const [savedGameArray, setSavedGameArray] = useState([])
    const [charState, dispatchCharState] = useReducer(charReducer, defaultCharState)
    const [gameState, dispatchGameState] = useReducer(gameReducer, defaultGameState)
    // Ongoing listeners:
    // User Account
    // gameList Array
    // Character Array
    // Saved Game Array
    useEffect(() => {

        // User Account listener
        onValue(ref(db, 'users/' + auth.currentUser.uid),
            (snapshot) => {
                // If there is a user account in the cloud at with this UID
                // copy it into local storage
                // and maintain listener status
                if (snapshot.exists()) {
                    dispatchUserState(updateUserState(snapshot.val()))
                } else {
                    // If no user account exists, clear local state
                    dispatchUserState(clearUserState())
                }
            })

        // gameList array listener
        onValue(ref(db, 'gameList'),
            (snapshot) => {
                const updatedArray = [];
                if (snapshot.exists()) {
                    snapshot.forEach((childSnapShot) => {
                        updatedArray.push(childSnapShot.val().gameID)
                    })
                    setGameArray(updatedArray)
                } else {
                    setGameArray([])
                }
            })

        // current user character array listener
        onValue(ref(db, 'characters/' + auth.currentUser.uid),
            (snapshot) => {
                const characterArray = [];
                if (snapshot.exists()) {
                    snapshot.forEach(childSnapShot => {
                        characterArray.push(childSnapShot.val())
                    })
                }
                setCharArray(characterArray)
            })

        // saved game array listener
        onValue(ref(db, 'savedGames/' + auth.currentUser.uid),
            (snapshot) => {
                const userSavedGames = [];
                if (snapshot.exists()) {
                    snapshot.forEach((savedGame) => {
                        userSavedGames.push(savedGame.val())
                    })
                }
                setSavedGameArray(userSavedGames)

            })

        return () => {
            // Remove the listeners when gameSetup is closed
            off(ref(db, 'users/' + auth.currentUser.uid))
            off(ref(db, 'gameList'))
            off(ref(db, 'characters/' + auth.currentUser.uid))
            off(ref(db, 'savedGames/' + auth.currentUser.uid))
        }
    }, [])

    // charState updater
    useEffect(() => {
        // Update the current character state if it changes
        // Clear local character state if no character state
        // exists in the cloud
        if (userState.currentCharacterID) {
            let charObject = charArray.find(character =>
                character.charID === userState.currentCharacterID)
            dispatchCharState(setCharState(charObject))
        } else {
            dispatchCharState(setNoCurrentChar())
        }
    }, [userState.currentCharacterID, charArray])

    // gameState listeners
    useEffect(() => {
        if (userState.gameID) {
            // When the local gameID changes, if the local gameID exists
            // start listeners to get the game information

            // Ongoing gameState Static listener
            onValue(ref(db, 'gameSetup/' + userState.gameID + '/static'),
                (snapshot) => {

                    if (snapshot.exists()) {
                        // If the game information exists, copy it to local
                        dispatchGameState(updateGameStatic(snapshot.val()))
                    } else {
                        // If the game information does not exist,
                        if (gameState.static.key === null) {
                            // check for a game key.
                            // Without a key, clear the local state
                            dispatchGameState(clearGameState())
                            // then close the listener
                            off(ref(db, 'gameSetup/' + userState.gameID + '/static'))
                        }
                    }
                })

            // Ongoing gameState Active listener
            onValue(ref(db, 'gameSetup/' + userState.gameID + '/active'),
                (snapshot) => {
                    if (snapshot.exists()) {
                        // If the game information exists, copy it to local
                        dispatchGameState(updateGameActive(snapshot.val()))
                        // console.log('gameState singles update: ', snapshot.val())
                    }
                })

            // Ongoing playerList listener
            onValue(ref(db, 'gameSetup/' + userState.gameID + '/playerList'),
                (snapshot) => {

                    if (snapshot.exists()) {
                        const playerList = [];
                        snapshot.forEach((player) => {
                            playerList.push(player.val());
                        })
                        const otherPlayers = playerList.filter(player => player.uid !== auth.currentUser.uid)
                        if (playerList.length > 0) {
                            dispatchGameState(updatePlayerList(otherPlayers))
                        }
                    }
                    else {
                        dispatchGameState(clearPlayerList())
                        off(ref(db, 'gameSetup/' + userState.gameID + '/playerList'))
                    }

                })

            // Ongoing classStorage listener
            onValue(ref(db, 'gameSetup/' + userState.gameID + '/classStorage'),
                snapshot => {
                    if (snapshot.exists()) {
                        const updatedClassArray = [];
                        snapshot.forEach((playerclass) => {
                            updatedClassArray.push(playerclass.val())
                        })
                        dispatchGameState(updateClassList(updatedClassArray))

                    } else {
                        dispatchGameState(clearClassList())
                        off(ref(db, 'gameSetup/' + userState.gameID + '/classStorage'))
                    }
                })

            // Ongoing readyCheck listener
            onValue(ref(db, 'gameSetup/' + userState.gameID + '/readyCheck'),
                snapshot => {
                    if (snapshot.exists()) {
                        const updatedReadyList = [];
                        snapshot.forEach((readyPlayer) => {
                            updatedReadyList.push(readyPlayer.val())
                        })
                        // Update local state with the list of ready players
                        dispatchGameState(updateReadyList(updatedReadyList))
                        // If this player is in the list, 
                        // update local state to reflect
                        if (updatedReadyList.includes(auth.currentUser.uid)) {
                            dispatchGameState(updateReadyStatus(true))
                        } else {
                            // otherwise set this player's ready status to false
                            dispatchGameState(updateReadyStatus(false))
                        }
                    } else {
                        // if there are no ready players, clear the ready list locally
                        dispatchGameState(clearReadyList())
                        // clear the current player's ready status
                        dispatchGameState(updateReadyStatus(false))
                    }

                })

        } else {
            dispatchGameState(clearGameState())
        }

        return () => {
            off(ref(db, 'gameSetup/' + userState.gameID + '/static'))
            off(ref(db, 'gameSetup/' + userState.gameID + '/active'))
            off(ref(db, 'gameSetup/' + userState.gameID + '/playerList'))
            off(ref(db, 'gameSetup/' + userState.gameID + '/classStorage'))
            off(ref(db, 'gameSetup/' + userState.gameID + '/readyCheck'))
        }
    }, [userState.gameID])

    // The presence of a local gameState.static.key will
    // confirm that a game has started, and move all players to 
    // the activeGame screen
    useEffect(() => {
        // If a game key exists, 
        // and the list of ready players is the same length as the list of (other) players
        // and THIS player is ready
        // navigate to the Introductions screen on the ActiveGame
        // as the game has begun

        if (gameState.static.key !== null &&
            gameState.readyList.length === (gameState.playerList.length + 1) &&
            gameState.active.ready) {
                console.log('ready to record current game', gameState)
            startRecordCurrentGame(
                auth.currentUser.uid, gameState.static.key, gameState.static.host
            )
            navigate('/activeGame/introductions')
            startRemoveGameCode(auth.currentUser.uid, gameState.static.gameID)
        }

    }, [gameState.static.key, gameState.readyList, gameState.active.ready])



    // Join an active game when a character has been selected and gameID entered
    useEffect(() => {

        // Add the current character class code to the active game 
        // and the UID and current character key to the list of 
        // players on the Active Game

        if (userState.currentCharacterID && userState.gameID) {
            startJoinActiveGame(
                auth.currentUser.uid,
                userState.gameID,
                userState.currentCharacterID,
                charState.classCode
            )
        }

    }, [charState.classCode, userState.gameID])



    return (
        <div>
            <AuthWrapper />
            <JoiningHosting
                userState={userState}
                dispatchGameState={dispatchGameState}
                gameArray={gameArray}
            />
            <ChallengeDisplay
                gameState={gameState}
            />
            <PlayingAs
                userState={userState}
                charState={charState}
            />
            <CharacterChallengeNavBar

            />
            <Routes>
                <Route index element={<GameInstructions />} />
                <Route
                    path="gameInstructions"
                    element={<GameInstructions />}
                />
                <Route
                    path='selectCharacter'
                    element={
                        <CharacterSelect
                            charState={charState}
                            charArray={charArray}
                        />
                    } />
                <Route
                    path='selectChallenges'
                    element={
                        <NewLoadWrapper
                            userState={userState}
                            gameState={gameState}
                            savedGameArray={savedGameArray}
                        />
                    } />
                <Route
                    path='restOfParty'
                    element={
                        <RestOfParty
                            gameState={gameState}
                        />
                    } />
            </Routes>
            <StartGame
                userState={userState}
                gameState={gameState}
                dispatchGameState={dispatchGameState}
            />
        </div>
    )

}

export { GameSetup as default }



