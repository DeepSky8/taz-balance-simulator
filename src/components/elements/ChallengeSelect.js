import React, { useEffect, useReducer, useState } from "react";
import {
    setVillainObject,
    setRelicObject,
    setLocationObject,
    toggleLocation,
    toggleRelic,
    toggleVillain,
    startSetChallenges
} from "../../actions/challengeSelectActions";
import { challengeSelectReducer, defaultChallengeState } from "../../reducers/challengeSelectReducer";
import ChallengeOptionsList from "./ChallengeOptionsList";
import { villainObjectsArray } from "./Challenges/mission-elements/m-villain";
import { relicObjectsArray } from "./Challenges/mission-elements/m-relic";
import { locationObjectsArray } from "./Challenges/mission-elements/m-location";

const ChallengeSelect = ({ setupState, dispatchSetupState }) => {
    const [challengeState, dispatchChallengeState] = useReducer(challengeSelectReducer, defaultChallengeState)
    const [challengesUpdated, updateChallenge] = useState(0)
    const dispatchedChallengeCode = () => {
        updateChallenge(challengesUpdated + 1)
    }
    const uid = setupState.uid
    const gameID = setupState.gameID

    const toggleVillainDisplay = () => {
        dispatchChallengeState(toggleVillain())
    }
    const toggleRelicDisplay = () => {
        dispatchChallengeState(toggleRelic())
    }
    const toggleLocationDisplay = () => {
        dispatchChallengeState(toggleLocation())
    }

    const villainDispatch = (challengeObject) => {
        if (!setupState.joiningGame) {

            // const challengeObject =
            //     challengeTransformer(villainObjectsArray, challengeCode)
            dispatchChallengeState(setVillainObject(challengeObject))
            dispatchedChallengeCode()
        }
    }

    const relicDispatch = (challengeObject) => {
        if (!setupState.joiningGame) {

            // const challengeObject =
            //     challengeTransformer(relicObjectsArray, challengeCode)
            dispatchChallengeState(setRelicObject(challengeObject))
            dispatchedChallengeCode()
        }
    }

    const locationDispatch = (challengeObject) => {
        if (!setupState.joiningGame) {

            // const challengeObject =
            //     challengeTransformer(locationObjectsArray, challengeCode)
            dispatchChallengeState(setLocationObject(challengeObject))
            dispatchedChallengeCode()
        }
    }

    useEffect(() => {
        startSetChallenges(
            {
                villainCode: challengeState.selectedVillainObject.challengeCode,
                relicCode: challengeState.selectedRelicObject.challengeCode,
                locationCode: challengeState.selectedLocationObject.challengeCode
            },
            uid
        )
    }, [challengesUpdated])

    return (
        <div>
            {!setupState.joiningGame &&
                <div>
                    Battle
                    <button onClick={toggleVillainDisplay}>
                        {challengeState.selectedVillainObject.challengeName}
                    </button>
                    to reclaim
                    <button onClick={toggleRelicDisplay}>
                        {challengeState.selectedRelicObject.challengeName}
                    </button>

                    {challengeState.selectedLocationObject.challengeCode === 'l3'
                        ?
                        'on'
                        :
                        'in'
                    }

                    <button onClick={toggleLocationDisplay}>
                        {challengeState.selectedLocationObject.challengeName}
                    </button>
                </div>
            }

            {setupState.joiningGame &&
                <div>
                    Your party will battle
                    <button onClick={toggleVillainDisplay}>
                        {challengeState.selectedVillainObject.challengeName}
                    </button>
                    to reclaim
                    <button onClick={toggleRelicDisplay}>
                        {challengeState.selectedRelicObject.challengeName}
                    </button>

                    {challengeState.selectedLocationObject.challengeCode === 'l3' ?
                        'on'
                        :
                        'in'
                    }

                    <button onClick={toggleLocationDisplay}>
                        {challengeState.selectedLocationObject.challengeName}
                    </button>
                </div>
            }


            <div>
                {challengeState.villainDisplayed &&
                    <ChallengeOptionsList
                        challengeObjectArray={villainObjectsArray}
                        challengeDispatch={villainDispatch}
                        joining={setupState.joiningGame}
                    />}

                {challengeState.relicDisplayed &&
                    <ChallengeOptionsList
                        challengeObjectArray={relicObjectsArray}
                        challengeDispatch={relicDispatch}
                        joining={setupState.joiningGame}
                    />}

                {challengeState.locationDisplayed &&
                    <ChallengeOptionsList
                        challengeObjectArray={locationObjectsArray}
                        challengeDispatch={locationDispatch}
                        joining={setupState.joiningGame}
                    />}
            </div>
        </div>
    )
}

export { ChallengeSelect as default }

// startSetChallenges(challengeCode, auth.currentUser.uid, )

