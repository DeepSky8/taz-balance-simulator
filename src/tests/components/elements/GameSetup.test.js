import React from "react";
import ReactShallowRenderer from 'react-test-renderer/shallow';
import { GameSetup } from "../../../components/elements/GameSetup";

test('should render GameSetup screen', () => {
    const renderer = new ReactShallowRenderer();
    renderer.render(<GameSetup />)
    expect(renderer.getRenderOutput()).toMatchSnapshot();
})