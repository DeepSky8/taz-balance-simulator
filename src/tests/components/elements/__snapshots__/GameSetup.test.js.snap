// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render GameSetup screen 1`] = `
<p>
  Game Setup page
</p>
`;
